### ISSUE TITLE
LTO generate undefined symbol

### ISSUE THREAD

**Issue by InnotSagg** [ldc-developers/ldc#3500]
this is the test code:

```d
@nogc nothrow:
struct A {
	@disable this(this);
}
extern(C) int main(){

        return 0;
}
```

build this code without lto:

```sh
ldc2 -betterC -c test.d 
nm test.o 
0000000000000000 T _D4test1A14__aggrPostblitMFNaNbNiNfZv
0000000000000000 T _D4test1A8opAssignMFNaNbNcNiNjNeSQBfQBdZQi
0000000000000000 T main
```

build this code with lto:

```sh
ldc2 -betterC -c test.d -flto=full
llvm-nm-10 ./test.o 
                 U _D4test1A10__postblitMFZv
---------------- T _D4test1A14__aggrPostblitMFNaNbNiNfZv
                 U _D4test1A15__fieldPostblitMFNaNbNiNfZv
---------------- T _D4test1A8opAssignMFNaNbNcNiNjNeSQBfQBdZQi
---------------- T main
```

this code will link successful normally,  but for MingGW target lld-link throw this error:

```sh
ldc2 -betterC test.d -flto=full
lld-link: error: undefined symbol: __D4test1A10__postblitMFZv
>>> referenced by test.d
>>>               test.o

lld-link: error: undefined symbol: __D4test1A15__fieldPostblitMFNaNbNiNfZv
>>> referenced by test.d
>>>               test.o
clang-10: error: linker command failed with exit code 1 (use -v to see invocation)
Error: /opt/d/.APP/prefix/wllvm86x/bin/clang failed with status: 1
```

build with lto also generate unnecessary __ctor on my project, but I am not able to minimal the test code.

Correct me if I am wrong, `__D4test1A10__postblitMFZv` symbol should not be created in this case.

**Comment by kinke** [comment]
Next time, please add new findings (also happening for MSVC) as a separate post/comment, otherwise we don't get notified. - As it doesn't happen for Linux targets (https://run.dlang.io/is/HKJ4vb) with basically equivalent IR generated by LDC, I consider this an LLD LTO issue. The 2 functions are declared in IR, but not referenced, so there's no obvious reason for any complaints.

**Comment by InnotSagg** [comment]
Thanks for explain, will add comment.

I think the issue here is should not create this symbol,  like NON-LTO version object.

with extra non-used symbol could slow down the link time.

**Comment by kinke** [comment]
> I think the issue here is should not create this symbol, like NON-LTO version object.

Sounds reasonable, but that'd be LLVM's job then, as the IR generated by LDC is identical.

**Comment by changloong** [comment]
> We could skip the useless IR declaration in LDC, but then again, if the user forward-declares some unused function, he/she will end up with the same problem with LTO on Windows...

Maybe for next release add warning about use forward-declares with  lld-link at `release Known issues list`.

`@disable this(this)` should not create any IR declaration, good to know we can fix it on LDC side.

**Comment by JohanEngelen** [comment]
yeah I think it is an LLVM bug to error on the final link for unreferenced undefined symbols.
Clang does not emit forward referenced functions in IR if they are not referenced, but I would not expect a frontend to do that optimization/culling.

**Comment by InnotSagg** [comment]
test case update:

```d
@nogc nothrow:
struct A {
        @disable this();
        @disable this(this);
}
extern(C) int main(){
        return 0;
}
```

3  undefined symbol: `_D4test1A6__ctorMFNcZSQuQr`, `_D4test1A10__postblitMFZv`, `_D4test1A15__fieldPostblitMFNaNbNiNfZv`

**Comment by kinke** [comment]
That's clear. This is the simplest variant, failing with both thin and full LTO for Win64:
```D
void foo();
void main() {}
```

**Comment by kinke** [comment]
Adding `-O` makes it work, and that's probably why this hasn't been fixed already.

**Comment by InnotSagg** [comment]
hi @kinke 

I try with `-O` get a new error, this is the reduced test case:

```d
import ldc.llvmasm;
import ldc.attributes;
@nogc nothrow:
__gshared int A = 1;
extern (C) void fiber_entryPoint() {
        A++;
}
void trampoline() @naked {
                    __asm(
                       `sub $$32, %rsp
                        call fiber_entryPoint
                        xor %rcx, %rcx
                        jmp *%rcx`,
                        "~{rsp},~{rcx}"
                    );
}
extern(C) int main(){
        trampoline();
        return A;
}
```

this build ok without `-flto=full` or `-flto=thin`.

`ldc2 test.d -betterC -flto=full -O -mtriple=x86_64-windows-msvc`  or with `-mtriple=x86_64-w64-mingw32` throw this error:

```sh
lld: error: undefined symbol: fiber_entryPoint
>>> referenced by lto.tmp:(main)
```

**Comment by InnotSagg** [comment]
window native build has the same error.

**Comment by kinke** [comment]
This is something completely different. If you look at the original definition in druntime (https://github.com/ldc-developers/druntime/blob/cc97ccd00d4082221eee1d5afdbd775201d75877/src/core/thread/fiber.d#L168), you'll see that I've explicitly added `@assumeUsed`, to prevent LLVM from not emitting the function (which is only referenced in inline asm).

**Comment by kinke** [comment]
Before you post any more LTO issues or problems, note that LDC itself is built with LTO (even across D and C++).

**Comment by InnotSagg** [comment]
Add `@ assumeUsed` fix the problem,  my app is able to build with LTO for window, reduce 20% size.

Thanks for the great work.
